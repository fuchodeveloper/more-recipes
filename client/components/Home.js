import React, { Component } from 'react';
import { render } from 'react-dom';
import { Link } from 'react-router-dom';
import noodles from '../assets/img/noodles.jpg';
import Header from '../components/navigation/Header';
import FlashMessagesList from './flash/FlashMessagesList';

class Home extends Component {
  render() {
    return (
      <div>
        {/* Header component for navigation */}
        <Header />
        
        
        {this.props.children}
        
          <div className="container margin-top-70">
          <FlashMessagesList />
          
              <h1 className="text-center p-4 center-hero-title">Awesome Recipes Just For You</h1>
              <FlashMessagesList />
              <form action="#" method="post">
              
                  <div className="input-group mt-2 mb-2 p-1">
                      <input type="text" className="form-control p-3" placeholder="Try: 'Jollof Rice' " aria-describedby="basic-addon2"/>
                      <input type="submit" value="SEARCH" className="btn btn-primary input-group-addon"/>
                  </div>
              </form>
              <div className="margin-top-50 margin-bottom-50"/>

              <div>
                <h3 className="popular-text">Most popular recipes</h3>
              </div>

              <div className="card-deck">
                <div className="card mt-1">
                    <img className="card-img-top" src={noodles} alt="Recipe image"/>
                    <div className="card-body">
                        <h4 className="card-title">Card title</h4>
                        <p className="card-text">This is a wider card with supporting text below as a natural lead-in to additional content. This content is a little bit longer.</p>
                        <div className="card-footer custom-card-footer-bg">
                            <p className="card-text"><small className="text-muted">30 <i className="fa fa-eye" aria-hidden="true"/> . 20 <i className="fa fa-thumbs-up" aria-hidden="true"/> . 42 <i className="fa fa-star" aria-hidden="true"/></small></p>
                            <Link to="/details" className="btn btn-primary btn-primary-color">View Recipe</Link>
                        </div>
                    </div>
                </div>
                <div className="card mt-1">
                <img className="card-img-top" src={noodles} alt="Recipe image"/>
                    <div className="card-body">
                        <h4 className="card-title">Card title</h4>
                        <p className="card-text">This is a wider card with supporting text below as a natural lead-in to additional content. This content is a little bit longer.</p>
                        <div className="card-footer custom-card-footer-bg">
                            <p className="card-text"><small className="text-muted">30 <i className="fa fa-eye" aria-hidden="true"/> . 20 <i className="fa fa-thumbs-up" aria-hidden="true"/> . 42 <i className="fa fa-star" aria-hidden="true"/></small></p>
                            <a href="template/details.html" className="btn btn-primary btn-primary-color">View Recipe</a>
                        </div>
                    </div>
                </div>

                <div className="card mt-1">
                <img className="card-img-top" src={noodles} alt="Recipe image"/>
                <div className="card-body">
                    <h4 className="card-title">Card title</h4>
                    <p className="card-text">This is a wider card with supporting text below as a natural lead-in to additional content. This content is a little bit longer.</p>
                    <div className="card-footer custom-card-footer-bg">
                        <p className="card-text"><small className="text-muted">30 <i className="fa fa-eye" aria-hidden="true"/> . 20 <i className="fa fa-thumbs-up" aria-hidden="true"/> . 42 <i className="fa fa-star" aria-hidden="true"/></small></p>
                        <a href="template/details.html" className="btn btn-primary btn-primary-color">View Recipe</a>
                    </div>
                </div>
              </div>
          </div>

          <div className="clearfix mt-4"/>
          {/* Bottom Navigation */}
          <nav aria-label="Page navigation example">
              <ul className="pagination justify-content-center">
                  <li className="page-item disabled">
                  <a className="page-link" href="#">Previous</a>
                  </li>
                  <li className="page-item"><a className="page-link" href="#">1</a></li>
                  <li className="page-item"><a className="page-link" href="#">2</a></li>
                  <li className="page-item"><a className="page-link" href="#">3</a></li>
                  <li className="page-item">
                  <a className="page-link" href="#">Next</a>
                  </li>
              </ul>
          </nav>

          </div>
      
      </div>
    );
  }
}

export default Home;