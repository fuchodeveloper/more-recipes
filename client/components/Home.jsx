import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ReactPaginate from 'react-paginate';
import isEmpty from 'lodash/isEmpty';
import AllRecipes from './recipes/AllRecipes';
import recipeSearch from '../action/recipes/recipeSearchAction';
import getAllRecipesAction from '../action/recipes/getAllRecipesAction';

/**
 * @description class to display all recipes
 *
 * @class Home
 *
 * @extends {Component}
 */
export class Home extends Component {
  /**
   * @description Creates an instance of Home.
   *
   * @param {Object} props
   *
   * @memberof Home
   */
  constructor(props) {
    super(props);
    this.state = {
      searchQuery: '',
      recipes: {},
      pageCount: ''
    }; // Initialize the state
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.onPageChange = this.onPageChange.bind(this);
  }

  /**
   * @description lifecycle method to show all recipes
   *
   * @memberof Home
   *
   * @returns {undefined}
   */
  componentDidMount() {
    this.props.recipesAction(this.state.pageCount);
  }

  /**
 * @description lifecycle method used to update state
 *
 * @param {Object} nextProps
 *
 * @memberof Home
 *
 * @returns {undefined}
 */
  componentWillReceiveProps(nextProps) {
    this.setState({
      recipes: nextProps.recipes,
      pageCount: nextProps.pageCount
    });
  }

  /**
 * @description function to handle focus on search
 *
 * @memberof Home
 * @returns {undefined}
 */
  onFocus() {
    this.context.router.history.push('/search');
  }

  /**
 * function to handle page number change
 *
 * @param {Number} current
 *
 * @memberof Home
 *
 * @returns {undefined}
 */
  onPageChange(current) {
    current.selected += 1;
    this.props.recipesAction(current.selected);
  }

  /**
 *
 *
 * @param {Object} event
 *
 * @memberof Home
 *
 * @return {undefined}
 */
  onChange(event) {
    event.preventDefault();

    this.setState({ [event.target.name]: event.target.value });
  }

  /**
 * @description handle form submit
 *
 * @param {Object} event
 *
 * @memberof Home
 *
 * @returns {undefined}
 */
  onSubmit(event) {
    event.preventDefault();
    this.props.recipeSearch(this.state.searchQuery);
  }

  /**
 * @description Render the JSX template
 *
 * @memberof Home
 *
 * @returns  {JSX} JSX representation of component
 */
  render() {
    const { isFetching } = this.state;

    if (isFetching) {
      return (
        <h2 className="text-center">Loading...</h2>
      );
    }


    if (isEmpty(this.state.recipes)) {
      return (
        <div>

          {this.props.children}

          <div className="container margin-top-70">

            <h1
              id="hero-title"
              className="text-center p-4 center-hero-title"
            >Awesome Recipes Just For You
            </h1>
            <div />
            <form onSubmit={this.onSubmit} >

              <div className="input-group mt-2 mb-2 p-1">

                <input
                  id="searchQuery"
                  type="text"
                  className="form-control p-3"
                  placeholder="Try: 'Jollof Rice' "
                  aria-describedby="searchQuery"
                  name="searchQuery"
                  value={this.state.searchQuery}
                  onFocus={this.onFocus}
                  required
                />

                <input
                  type="submit"
                  value="SEARCH"
                  className="btn btn-primary input-group-addon"
                />
              </div>
            </form>
            <div className="margin-top-50 margin-bottom-50" />

            <div>
              <h3
                className="popular-text"
                id="recent"
              >Most recent recipes
              </h3>
            </div>

            <div>
              <div className="text-center">
                <i className="recipe-title">No recipes available yet</i>

              </div>

            </div>

            <div className="clearfix mt-4" />

          </div>


          <div className="clearfix m-5" />


        </div>
      );
    }

    return (
      <div>
        {this.props.children}

        <div className="container margin-top-70">

          <h1
            className="text-center p-4 center-hero-title"
          >Awesome Recipes Just For You
          </h1>
          <div />
          <form onSubmit={this.onSubmit} >

            <div className="input-group mt-2 mb-2 p-1">

              <input
                id="searchQuery"
                type="text"
                className="form-control p-3"
                placeholder="Try: 'Jollof Rice' "
                aria-describedby="basic-addon2"
                name="searchQuery"
                value={this.state.searchQuery}
                onFocus={this.onFocus}
                required
              />

              <input
                type="submit"
                value="SEARCH"
                className="btn btn-primary input-group-addon"
              />
            </div>
          </form>
          <div className="margin-top-50 margin-bottom-50" />

          <div>
            <h3 className="popular-text">Most recent recipes</h3>
          </div>

          <div>
            <div className="row">
              {
                Object
                  .keys(this.state.recipes)
                  .map(key => (<AllRecipes
                    key={key}
                    details={this.state.recipes[key]}
                  />))
                }

            </div>

          </div>

          <div className="clearfix mt-4" />

          <div className="row">
            <ReactPaginate
              pageCount={parseInt(this.state.pageCount, 10)}
              pageRangeDisplayed={5}
              marginPagesDisplayed={3}
              previousLabel="Previous"
              nextLabel="Next"
              breakClassName="text-center"
              initialPage={0}
              containerClassName="container pagination justify-content-center"
              pageClassName="page-item"
              pageLinkClassName="page-link"
              activeClassName="page-item active"
              previousClassName="page-item"
              nextClassName="page-item"
              nextLinkClassName="page-link"
              previousLinkClassName="page-link"
              onPageChange={this.onPageChange}
            />
          </div>

        </div>


        <div className="clearfix m-5 mb-5" />
      </div>
    );
  }
}

Home.contextTypes = {
  router: PropTypes.object
};

const mapStateToProps = state => ({
  recipes: state.recipesReducer.recipes,
  isFetching: state.isFetching,
  searchResult: state.searchResult,
  pageCount: state.recipesReducer.pageCount
});

const mapDispatchToProps = dispatch => ({
  recipesAction: pageCount => dispatch(getAllRecipesAction(pageCount)),
  recipeSearch: searchContent => dispatch(recipeSearch(searchContent))
});

Home.defaultProps = {
  recipes: [],
  children: null,
  pageCount: 1
};

Home.propTypes = {
  recipesAction: PropTypes.func.isRequired,
  recipeSearch: PropTypes.func.isRequired,
  recipes: PropTypes.arrayOf(PropTypes.shape),
  pageCount: PropTypes.number,
  children: PropTypes.node
};

export default connect(mapStateToProps, mapDispatchToProps)(Home);
