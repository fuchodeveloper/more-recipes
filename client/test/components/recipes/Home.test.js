/* eslint-disable max-len */
import React from 'react';
import { configure, shallow } from 'enzyme';
import sinon from 'sinon';
import Adapter from 'enzyme-adapter-react-16';
import expect from 'expect';
import { Home } from '../../../components/Home';

configure({ adapter: new Adapter() });

let props;

const setup = () => {
  props = {
    recipes: [],
    recipesAction: jest.fn(),
    recipeSearch: jest.fn(),
    router: {
      history: {
        push: '/'
      }
    }
  };
  return shallow(<Home {...props} />);
};


describe('Home component', () => {
  const page = {
    selected: 1
  };

  describe('Home component', () => {
    it('should render with right amount of elements', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });

  it(
    'should call componentDidMount lifecycle method on Home component',
    () => {
      const spy = sinon.spy(Home.prototype, 'componentDidMount');
      shallow(<Home{...props} componentDidMount={spy} />)
        .instance().componentDidMount({ setState: () => 1 });
    }
  );

  it('should call the onPageChange handler', () => {
    const spy = sinon.spy(Home.prototype, 'onPageChange');
    shallow(<Home {...props} onPageChange={spy} />)
      .instance().onPageChange(page);
  });

  it('component should receive props', () => {
    const wrapper = setup();
    const action = wrapper.instance();
    const spy = sinon.spy(
      Home.prototype,
      'componentWillReceiveProps'
    );
    const nextProps = {
      recipes: [
        {
          id: 1,
          userId: 1,
          name: 'jollof rice and beans',
          favoriteCount: 0,
          ingredients: 'rice and beans',
          direction: 'proin eget tortor risus. nulla quis lorem ut libero malesuada feugiat. curabitur non nulla  sit amet nisl tempus convallis quis ac lectus. curabitur arcu erat,  accumsan id imperdiet et, porttitor at sem. cras ultricies ligula sed magna dictum porta.',
          image: 'https://res.cloudinary.com/fuchodeveloper/image/upload/v1516760699/noodles_c6ltkq.jpg',
          views: 59,
          upVotes: 0,
          downVotes: 0,
          recipeOwnerView: true,
          createdAt: '2018-02-04T16:52:28.768Z',
          updatedAt: '2018-02-06T02:55:06.389Z'
        }
      ],
      pageCount: 1
    };
    shallow(<Home {...props} componentWillReceiveProps={spy} />);
    action.componentWillReceiveProps(nextProps);
    expect(action.state.recipes).toBe(nextProps.recipes);
    expect(action.state.pageCount).toBe(nextProps.pageCount);
  });
});
