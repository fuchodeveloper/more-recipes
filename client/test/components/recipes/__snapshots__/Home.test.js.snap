// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component Home component should render with right amount of elements 1`] = `undefined`;
